### PR TITLE
intersphinx: fix flipped use of `intersphinx_cache_limit`.

### PR DESCRIPTION
**Pull request by broken-pen**
Subject: intersphinx: fix flipped use of `intersphinx_cache_limit`.

### Feature or Bugfix
- Bugfix

### Purpose
Follow-up to #12514, which fixed one bug surrounding [`intersphinx_cache_limit`](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#confval-intersphinx_cache_limit), but introduced another by flipping a comparison. 

### Detail
The previous patch properly took negative cache limits into account ([which are meant to always keep the cache](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#confval-intersphinx_cache_limit)). However, in the process, it **flipped the comparison** between `intersphinx_cache_limit` and zero. The consequence was that the cache was always kept for positive values, and always refreshed for negative values.

The provided unit test **didn't catch the mistake** because it merely checked the return value of `_fetch_inventory_group()`, which is True if the cache has been updated *successfully*. However, what should've been tested was whether we *tried* to update the cache.

The problem was that the test attempted to fetch the remote inventory from a nonsense address, which is reasonable for a test. However, **the fetch fails, as expected**. And correctly Sphinx tells us that the cache wasn't updated, even though it *tried to*.

The new test **replaces the inner fetch function with a mock** that always returns some data. This means that "did we try to update the cache" and "did we successfullt update the cache" are now the same. It also lets us query the mock directly whether it has been called or not.

In addition, the **test also has been parametrized** over multiple values. This way, we test not only the first of the following cases of interest, but all three:

1. negative cache limit
2. positive cache limit, cache fresh
3. positive cache limit, cache stale

### Relates
- https://github.com/sphinx-doc/sphinx/pull/12514
- https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#confval-intersphinx_cache_limit

### CODE REVIEW

File: tests/test_extensions/test_ext_intersphinx.py [hunk 1]
```diff
@@ -745,30 +745,51 @@ def test_intersphinx_role(app):
...
-        # no need to read from remote
-        assert not _fetch_inventory_group(
+        updated = _fetch_inventory_group(
             project=project,
             cache=intersphinx_cache,
             now=now,
             config=app.config,
             srcdir=app.srcdir,
         )
+        # If we hadn't mocked `_fetch_inventory`, it would've made
```
Comment (picnixz) [inline tests/test_extensions/test_ext_intersphinx.py:790]: One way to make it work is to have a fake HTTP server serving the inventory as we usually do. It's my bad that I didn't think of it when I said that the test was fine.

**Review by picnixz — COMMENTED** [review]
I'm extremely sorry for the inconvenience. I should stop reviewing PRs and accepting them when I'm still sleepy or too tired. By the way, thank you for the tests (I *am* now tired so I'll still ask @AA-Turner to have a final check).

File: tests/test_extensions/test_ext_intersphinx.py (hunk 1 above)
Comment (picnixz) [reply tests/test_extensions/test_ext_intersphinx.py:790]: @AA-Turner Do you have a preference for this test or not? we can either create a fake HTTP server for the test or mock the tests like this. With an HTTP server, I feel that we wouldn't rely on possible changes of `_fetch_inventory` for whatever reason.

**Review by AA-Turner — APPROVED** [review]
Thank you Nico (@troiganto) for the excellent description and PR, sorry it took too long to review.

A

**Comment by troiganto** [comment]
np, thanks for the maintenance work! contributing was a lot simpler than for other projects, so I actually had a great time solving this!

**Comment by SilverRainZ** [comment]
I am sorry for introducing another bug when fixing a bug. Thanks for fixing this, I learned more unit test tips from this PR.

**Pull request merged**